### PR TITLE
Add is_pareto_efficient

### PR DESCRIPTION
**Pull request by QBatista**
Adds a function to determine if an action profile is Pareto efficient.

### CODE REVIEW

**Comment by oyamad** [comment]
For tests, better to include corner cases, where there are two payoff vectors that are identical and those in which the payoffs are equal only for one player.

**Comment by oyamad** [comment]
Should be rebased on `master`.

File: src/normal_form_game.jl [hunk 1]
```diff
@@ -612,3 +612,50 @@ function pure2mixed(num_actions::Integer, action::PureAction)
     mixed_action[action] = 1
     return mixed_action
 end
+
+# is_pareto_efficient & is_pareto_dominant
+
+for args = ((:is_pareto_efficient, &), (:is_pareto_dominant, |))
```
Comment (oyamad) [inline src/normal_form_game.jl]: I think `&&` and `||` should be used rather than the bitwise operators `&` and `|`, while I see `&&` and `||` do not work with `@eval`. What about defining functions:
```jl
_and(a, b) = a && b
_or(a, b) = a || b
```

Comment (oyamad) [inline src/normal_form_game.jl]: A variable name should be more descriptive here. For example use `f` for a function and maybe `op` for a logical operator?

File: src/normal_form_game.jl [hunk 2]
```diff
@@ -612,3 +612,50 @@ function pure2mixed(num_actions::Integer, action::PureAction)
...
+    @eval function $(args[1])(g::NormalFormGame,
+                     action_profile::PureActionProfile)
+            payoff_profile0 = g[action_profile...]
+            for action_profiles in CartesianRange(g.nums_actions)
```
Comment (oyamad) [inline src/normal_form_game.jl]: The plural `action_profiles` sounds strange.

File: test/test_normal_form_game.jl [hunk 3]
```diff
@@ -180,4 +180,65 @@
...
+        coordination_game_matrix = [4 0;
+                                    3 2]
+        matching_pennies_bimatrix = Array{Float64}(2, 2, 2)
+        matching_pennies_bimatrix[1, 1, :] = [1, -1]
+        matching_pennies_bimatrix[1, 2, :] = [-1, 1]
+        matching_pennies_bimatrix[2, 1, :] = [-1, 1]
+        matching_pennies_bimatrix[2, 2, :] = [1, -1]
+
+        games_dict = [NormalFormGame(coordination_game_matrix),
+                      NormalFormGame(matching_pennies_bimatrix)]
```
Comment (oyamad) [inline test/test_normal_form_game.jl]: Avoid repeating the same code. `@testset` can be nested. What about:

```jl
@testset "Tests on Pareto efficiency and dominance" begin
    # Define games
    @testset "Testing is_pareto_efficient" begin
        # Test is_pareto_efficient
    end
    @testset "Testing Pareto dominance" begin
        # Test is_pareto_dominant
    end
end
```

File: test/test_normal_form_game.jl [hunk 4]
```diff
@@ -180,4 +180,65 @@
...
+    @testset "Pareto efficiency with 2 players" begin
+        coordination_game_matrix = [4 0;
+                                    3 2]
+        equal_payoff_p1_bimatrix = Array{Float64}(2, 2, 2)
+        equal_payoff_p1_bimatrix[1, 1, :] = [1, -1]
+        equal_payoff_p1_bimatrix[1, 2, :] = [1, 1]
+        equal_payoff_p1_bimatrix[2, 1, :] = [1, 1]
+        equal_payoff_p1_bimatrix[2, 2, :] = [1, -1]
+        two_equal_payoffs_bimatrix = [2 0;
+                                      0 2]
```
Comment (oyamad) [inline test/test_normal_form_game.jl]: Make this a 3-player game?

File: src/normal_form_game.jl (hunk 1 above)
Comment (oyamad) [reply src/normal_form_game.jl]: This is not a good idea...
Your original idea seems to be the way to go. What about:

```jl
function pareto_inferior_to(payoff_profile1, payoff_profile2)
    all(payoff_profile2 .>= payoff_profile1) &&
    any(payoff_profile2 .> payoff_profile1)
end

function not_pareto_superior_to(payoff_profile1, payoff_profile2)
    any(payoff_profile2 .> payoff_profile1) ||
    all(payoff_profile2 .== payoff_profile1)
end
```

Comment (QBatista) [reply src/normal_form_game.jl]: Just to be clear, I should get rid of `is_pareto_efficient` and `is_pareto_dominant` and replace them by `pareto_inferior_to` and `not_pareto_superior_to` (including the tests)?

Comment (oyamad) [reply src/normal_form_game.jl]: No. I was referring to the operations in the `if` clause.

**Comment by oyamad** [comment]
Thanks!
